### PR TITLE
[query] Implement mqtt-hybrid  @open sesame 07/20 18:32

### PR DESCRIPTION
**Pull request by gichan-jang**
Implement mqtt-hybrid using nnstreamer-edge lib.
    
 @todo Update mqtt-hybrid test

``` 
# server1
gst-launch-1.0 tensor_query_serversrc id=12345 host=tcp://localhost port=1883 topic=tempTopic  ! \
other/tensors,num_tensors=1,dimensions=3:300:300:1,types=uint8,framerate=30/1 ! \
tensor_query_serversink async=false id=12345
```

``` 
# server2
gst-launch-1.0 tensor_query_serversrc id=54321 host=tcp://localhost port=1883 topic=tempTopic  ! \
other/tensors,num_tensors=1,dimensions=3:300:300:1,types=uint8,framerate=30/1 ! \
tensor_query_serversink async=false id=54321
```

```
# client
gst-launch-1.0 v4l2src ! videoconvert ! videoscale !  video/x-raw,width=300,height=300,format=RGB,framerate=30/1 ! \
tensor_converter ! tensor_query_client host=tcp://localhost port=1883 topic=tempTopic ! \
tensor_decoder mode=direct_video ! videoconvert ! ximagesink
```

**Self evaluation:**
1. Build test: [ *]Passed [ ]Failed [ ]Skipped
2. Run test: [ *]Passed [ ]Failed [ ]Skipped

### CODE REVIEW

**Comment by gichan-jang** [comment]
Indent checkers work weird..

File: gst/nnstreamer/tensor_query/nnstreamer_edge_mqtt.c [hunk 1]
```diff
@@ -338,3 +390,31 @@ nns_edge_mqtt_subscribe (nns_edge_h edge_h)
...
+nns_edge_mqtt_is_connected (nns_edge_h edge_h)
+{
+  nns_edge_handle_s *eh;
+  MQTTAsync handle;
+
+  eh = (nns_edge_handle_s *) edge_h;
+
+  if (!NNS_EDGE_MAGIC_IS_VALID (eh)) {
+    nns_edge_loge ("Invalid param, given edge handle is invalid.");
+    return NNS_EDGE_ERROR_INVALID_PARAMETER;
```
Comment (wooksong) [inline gst/nnstreamer/tensor_query/nnstreamer_edge_mqtt.c:407]: This function is supposed to return a boolean value.

File: gst/nnstreamer/tensor_query/nnstreamer_edge_internal.c [hunk 2]
```diff
@@ -660,7 +663,7 @@ _nns_edge_connect_to (nns_edge_handle_s * eh, int64_t client_id,
       /* Send ip and port to destination. */
       _nns_edge_cmd_init (&cmd, _NNS_EDGE_CMD_HOST_INFO, client_id);
 
-      _get_host_str (eh->recv_ip, eh->recv_port, &host);
+      _get_host_str (eh->listner_ip, eh->listner_port, &host);
```
Comment (jaeyun-jung) [inline gst/nnstreamer/tensor_query/nnstreamer_edge_internal.c:666]: typo? listner

File: gst/nnstreamer/tensor_query/nnstreamer_edge_internal.c [hunk 3]
```diff
@@ -1015,21 +1052,53 @@ nns_edge_start (nns_edge_h edge_h, bool is_server)
...
+  /** In the case of the server, let's publish server information. */
   eh->is_server = is_server;
-  if (!is_server && 0 == eh->recv_port) {
-    eh->recv_port = _get_available_port ();
-    if (eh->recv_port <= 0) {
-      nns_edge_loge ("Failed to start edge. Cannot get available port.");
-      nns_edge_unlock (eh);
-      return NNS_EDGE_ERROR_CONNECTION_FAILURE;
+  if (eh->is_server) {
+    if (0 != g_ascii_strcasecmp (eh->topic, "TCP_DIRECT")) {
```
Comment (jaeyun-jung) [inline gst/nnstreamer/tensor_query/nnstreamer_edge_internal.c:1061]: discussion: topic or connection protocol
IMO connection protocol is correct way (tcp-direct or mybrid)

File: gst/nnstreamer/tensor_query/nnstreamer_edge_internal.c [hunk 4]
```diff
@@ -1088,9 +1164,15 @@ nns_edge_release_handle (nns_edge_h edge_h)
   SAFE_FREE (eh->id);
   SAFE_FREE (eh->topic);
   SAFE_FREE (eh->ip);
-  SAFE_FREE (eh->recv_ip);
+  SAFE_FREE (eh->listner_ip);
```
Comment (jaeyun-jung) [inline gst/nnstreamer/tensor_query/nnstreamer_edge_internal.c:1167]: Q: are IP and listener-IP different?

File: gst/nnstreamer/tensor_query/nnstreamer_edge_mqtt.c [hunk 5]
```diff
@@ -181,6 +211,12 @@ nns_edge_mqtt_connect (nns_edge_h edge_h)
   nns_edge_logi ("Trying to connect MQTT (ID:%s, URL:%s:%d).",
       eh->id, eh->ip, eh->port);
 
+  g_cond_init (&eh->mqtt_gcond);
+  g_mutex_init (&eh->mqtt_mutex);
+  g_mutex_lock (&eh->mqtt_mutex);
```
Comment (jaeyun-jung) [inline gst/nnstreamer/tensor_query/nnstreamer_edge_mqtt.c:216]: line216 and 218, IMO lock is unnecessary here.

File: gst/nnstreamer/tensor_query/nnstreamer-edge-internal.h [hunk 6]
```diff
@@ -42,13 +42,17 @@ typedef struct {
...
-  char *recv_ip;
-  int recv_port;
+  char *listner_ip;
+  int listner_port;
   GHashTable *conn_table;
   GSocketListener *listener;
 
   /* MQTT */
   void *mqtt_handle;
+  GAsyncQueue *server_list;
```
Comment (jaeyun-jung) [inline gst/nnstreamer/tensor_query/nnstreamer-edge-internal.h:52]: how about adding new struct for mqtt?
Suggestion: nns_edge_mqtt_state_s in -mqtt.c / nns_edge_broker_handle in common.h

Comment (jaeyun-jung) [reply gst/nnstreamer/tensor_query/nnstreamer-edge-internal.h:52]: Or.. will you update these on nnstreamer-edge repo?

Comment (gichan-jang) [reply gst/nnstreamer/tensor_query/nnstreamer-edge-internal.h:52]: Ok, I will move mqtt data structure to -mqtt.c.

File: gst/nnstreamer/tensor_query/tensor_query_client.c [hunk 7]
```diff
@@ -235,7 +191,11 @@ gst_tensor_query_client_finalize (GObject * object)
   }
   g_async_queue_unref (self->msg_queue);
 
-  tensor_query_hybrid_close (&self->hybrid_info);
+  if (self->msg_queue) {
+    g_async_queue_unref (self->msg_queue);
```
Comment (jaeyun-jung) [inline gst/nnstreamer/tensor_query/tensor_query_client.c:195]: line 192, duplicated

**Review by jaeyun-jung — APPROVED** [review]
LGTM. Now we are migrating edge library to nnstreamer-edge repo. 
After merging this, let's discuss interfaces of edge library and implement it on both nnstreamer and edge repo before removing edge lib in this repo.

**Review by wooksong — APPROVED** [review]
LGTM

**Pull request merged**